course6/lec6_script_en.md

```diff
@@ -0,0 +1,347 @@
+
+![](../pics/queue_push.drawio.png)
+
+![](../pics/queue_push_more.drawio.png)
+
+![](../pics/queue_pop.drawio.png)
+
+Let's look at a specific example. Initially, we have an empty queue, so both stacks are empty. After one addition, we add a number to `back`. Then we organize the queue and find that the queue is not empty, but `front` is empty, which does not meet our previously stated invariant, so we rotate the stack `back` and move rotated elements to `front`. Afterwards, we continue to add elements to `back`. Since `front` is not empty, it meets the invariant, and we do not need additional processing.
+
+After that, our repeatedly additions are only the quick addition of new elements in `back`. Then, we remove elements from `front`. We check the invariant after the operation. We find that the queue is not empty, but `front` is empty, so we do retate `back` and move elements to `front` again. After that, we can normally take elements from `front`. 
+
+You can see that one rotation supports multiple removal operations, therefore the overall cost is much less than rebuilding the list every time.
+
+```moonbit no-check
+struct Queue[T] {
+  front: Stack[T]
+  back: Stack[T]
+}
+fn Queue::empty[T]() -> Queue[T] { {front: Empty, back: Empty} }
+
+// Store element at the end of the queue
+fn push[T](self: Queue[T], value: T) -> Queue[T] { 
+  normalize({ ..self, back: self.back.push(value)}) // By defining the first argument as self, we can use xxx.f()
+}
+
+// Remove the first element
+fn pop[T](self: Queue[T]) -> (Option[T], Queue[T]) { 
+  match self.front {
+    Empty => (None, self)
+    NonEmpty(top, rest) => (Some(top), normalize({ ..self, front: rest}))
+  }
+}
+
+// If front is empty, reverse back to front
+fn normalize[T](self: Queue[T]) -> Queue[T] { 
+  match self.front {
+    Empty => { front: self.back.reverse(), back: Empty }
+    _ => self
+  }
+}
+
+// Helper function: reverse the stack
+fn reverse[T](self: Stack[T]) -> Stack[T] { 
+  let mut result: Stack[T] = Empty
+  let mut current: Stack[T] = self
+  while let (Some(top), rest) = pop(current) {
```

Comment:
Are you sure you're writing MoonBit instead of Rust? Have you at least tried on <https://try.moonbitlang.cn>?

Reply:
How about

```moonbit
fn go(acc, xs: Stack[T]) {
  match xs {
    Empty => acc
    NonEmpty(top, rest) => go((NonEmpty(top, acc) : Stack[T]), rest)
  }
}

go(Empty, self)
```

Reply:
Yes, that will work